### PR TITLE
Add all available parameters to the specific ListOpts struct

### LINKED ISSUE #106: Pass more parameters for images
Hello,

it would be nice if I could pass parameters like `type` to the [`images` endpoint](https://docs.hetzner.cloud/#images-get-all-images).
I guess to accomplish this, the [ImageListOpts](https://github.com/hetznercloud/hcloud-go/blob/master/hcloud/image.go#L122) must be extended.

### PR DESCRIPTION
**Pull request by LKaemmerling**
Refactor all GetByName-Methods to use the new ListOpts parameters and the List function internally, no API change!

Closes #106

### CODE REVIEW

File: hcloud/action.go [hunk 1]
```diff
@@ -95,11 +96,28 @@ func (c *ActionClient) GetByID(ctx context.Context, id int) (*Action, *Response,
 // ActionListOpts specifies options for listing actions.
 type ActionListOpts struct {
 	ListOpts
+	Status []string
+	Sort   []string
+}
+
+func valuesForActionListOpts(opts ActionListOpts) url.Values {
```
Comment (thcyron) [inline hcloud/action.go]: What about making this a method `Values()` of `ActionListOpts`?

File: hcloud/action.go [hunk 2]
```diff
@@ -95,11 +96,28 @@ func (c *ActionClient) GetByID(ctx context.Context, id int) (*Action, *Response,
...
+	vals := valuesForListOpts(opts.ListOpts)
+	if len(opts.Status) > 0 {
```
Comment (thcyron) [inline hcloud/action.go]: No needed since you can range over empty and nil slices.

File: hcloud/action_test.go [hunk 3]
```diff
@@ -74,6 +74,34 @@ func TestActionClientList(t *testing.T) {
...
+				t.Errorf("expected status[0] to be running; got %q", status[0])
+			}
+			if status[1] != "error" {
+				t.Errorf("expected status[0] to be error; got %q", status[1])
+			}
+		}
+
+		sort := r.Form["sort"]
+		if len(sort) != 3 {
+			t.Errorf("expected status to contain 3 elements; got %q", sort)
```
Comment (thcyron) [inline hcloud/action_test.go]: typo: expected **status**

File: hcloud/action_test.go [hunk 4]
```diff
@@ -74,6 +74,34 @@ func TestActionClientList(t *testing.T) {
...
+		r.ParseForm()
+		status := r.Form["status"]
+		if len(status) != 2 {
+			t.Errorf("expected status to contain 2 elements; got %q", status)
+		} else {
+			if status[0] != "running" {
+				t.Errorf("expected status[0] to be running; got %q", status[0])
+			}
+			if status[1] != "error" {
+				t.Errorf("expected status[0] to be error; got %q", status[1])
```
Comment (thcyron) [inline hcloud/action_test.go]: Wrong index

File: hcloud/image.go [hunk 5]
```diff
@@ -91,22 +91,12 @@ func (c *ImageClient) GetByID(ctx context.Context, id int) (*Image, *Response, e
...
-	}
-
-	var body schema.ImageListResponse
-	resp, err := c.client.Do(req, &body)
-	if err != nil {
-		return nil, nil, err
+	images, response, err := c.List(ctx, ImageListOpts{Name: name})
+	var image *Image
+	if len(images) > 0 {
+		image = images[0]
```
Comment (thcyron) [inline hcloud/image.go]: If you’d check for `len(images) == 0` and return then immediately, you could skip the temporary variable and make the code easier to read (at least to me).

File: hcloud/action_test.go [hunk 6]
```diff
@@ -74,6 +74,34 @@ func TestActionClientList(t *testing.T) {
 		if perPage := r.URL.Query().Get("per_page"); perPage != "50" {
 			t.Errorf("expected per_page 50; got %q", perPage)
 		}
+
+		r.ParseForm()
+		status := r.Form["status"]
```
Comment (thcyron) [inline hcloud/action_test.go]: You can access query string parameters easier:

```go
r.URL.Query().Get("status")
```

File: hcloud/image.go [hunk 7]
```diff
@@ -121,11 +110,28 @@ func (c *ImageClient) Get(ctx context.Context, idOrName string) (*Image, *Respon
 // ImageListOpts specifies options for listing images.
 type ImageListOpts struct {
 	ListOpts
+	Type    []string
```
Comment (thcyron) [inline hcloud/image.go]:
```suggestion
	Type    []ImageType
```

File: hcloud/image.go [hunk 8]
```diff
@@ -121,11 +110,28 @@ func (c *ImageClient) Get(ctx context.Context, idOrName string) (*Image, *Respon
...
+	BoundTo *Server
+	Name    string
+}
```
Comment (thcyron) [inline hcloud/image.go]: `Sort` is missing.

File: hcloud/iso.go [hunk 9]
```diff
@@ -60,22 +60,11 @@ func (c *ISOClient) GetByID(ctx context.Context, id int) (*ISO, *Response, error
...
-	var body schema.ISOListResponse
-	resp, err := c.client.Do(req, &body)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	if len(body.ISOs) == 0 {
-		return nil, resp, nil
+	isos, response, err := c.List(ctx, ISOListOpts{Name: name})
+	if len(isos) > 0 {
```
Comment (thcyron) [inline hcloud/iso.go]: Sometimes it’s `if len(...) == 0`, sometime it’s `if len(...) > 0`. Should be consistent across all clients.

File: hcloud/action_test.go (hunk 6 above)
Comment (LKaemmerling) [reply hcloud/action_test.go]: This does not work with parameters which can be specified multiple times:
https://play.golang.org/p/N40PisXfxa9

File: hcloud/iso.go [hunk 10]
```diff
@@ -89,11 +78,20 @@ func (c *ISOClient) Get(ctx context.Context, idOrName string) (*ISO, *Response,
 // ISOListOpts specifies options for listing isos.
 type ISOListOpts struct {
 	ListOpts
+	Name string
+}
+
+func valuesForISOListOpts(opts ISOListOpts) url.Values {
```
Comment (thcyron) [inline hcloud/iso.go]: Should also be converted to `values()` methods.

File: hcloud/action_test.go (hunk 6 above)
Comment (thcyron) [reply hcloud/action_test.go]: Oh, I see.

> To access multiple values, use the map directly.

https://golang.org/pkg/net/url/#Values.Get

File: hcloud/ssh_key.go [hunk 11]
```diff
@@ -46,42 +46,21 @@ func (c *SSHKeyClient) GetByID(ctx context.Context, id int) (*SSHKey, *Response,
...
 }
 
 // GetByFingerprint retreives a SSH key by its fingerprint. If the SSH key does not exist, nil is returned.
 func (c *SSHKeyClient) GetByFingerprint(ctx context.Context, fingerprint string) (*SSHKey, *Response, error) {
-	path := "/ssh_keys?fingerprint=" + url.QueryEscape(fingerprint)
-	req, err := c.client.NewRequest(ctx, "GET", path, nil)
-	if err != nil {
-		return nil, nil, err
+	sshKeys, response, err := c.List(ctx, SSHKeyListOpts{Fingerprint: fingerprint})
+	var sshKey *SSHKey
```
Comment (thcyron) [inline hcloud/ssh_key.go]: Should adapt the same style without intermediate variable.

**Comment by thcyron** [comment]
```go
func valuesForListOpts(opts ListOpts) url.Values
```

Should also be converted to `func (l ListOpts) values() url.Values` I think.

File: hcloud/action.go [hunk 12]
```diff
@@ -95,11 +96,24 @@ func (c *ActionClient) GetByID(ctx context.Context, id int) (*Action, *Response,
 // ActionListOpts specifies options for listing actions.
 type ActionListOpts struct {
 	ListOpts
+	Status []string
```
Comment (thcyron) [inline hcloud/action.go]:
```suggestion
	Status []ActionStatus
```

File: hcloud/image.go [hunk 13]
```diff
@@ -121,11 +110,32 @@ func (c *ImageClient) Get(ctx context.Context, idOrName string) (*Image, *Respon
...
 	ListOpts
+	Type    []ImageType
+	BoundTo *Server
+	Name    string
+	Sort    []string
+}
+
+func (l ImageListOpts) values() url.Values {
+	vals := l.ListOpts.values()
+	for _, Type := range l.Type {
```
Comment (thcyron) [inline hcloud/image.go]:
```suggestion
	for _, typ := range l.Type {
```

File: hcloud/image.go [hunk 14]
```diff
@@ -121,11 +110,32 @@ func (c *ImageClient) Get(ctx context.Context, idOrName string) (*Image, *Respon
...
+		vals.Add("type", string(Type))
+	}
+	if l.BoundTo != nil {
+		vals.Add("bound_to", strconv.Itoa(l.BoundTo.ID))
+	}
+	if l.Name != "" {
+		vals.Add("name", l.Name)
+	}
+	for _, Sort := range l.Sort {
```
Comment (thcyron) [inline hcloud/image.go]:
```suggestion
	for _, sort := range l.Sort {
```

**Pull request merged**